cmd/zdb/zdb.c

```diff
@@ -796,38 +831,125 @@ dump_metaslab(metaslab_t *msp)
 static void
 print_vdev_metaslab_header(vdev_t *vd)
 {
-	(void) printf("\tvdev %10llu\n\t%-10s%5llu   %-19s   %-15s   %-10s\n",
-	    (u_longlong_t)vd->vdev_id,
+	char class[32] = "(";
+
+	if (vd->vdev_islog || vd->vdev_mg->mg_class[SPA_CLASS_LOG] != NULL)
+		(void) strcat(class, VDEV_CLASS_LOG",");
+	if (vd->vdev_mg->mg_class[SPA_CLASS_MOS] != NULL)
+		(void) strcat(class, VDEV_CLASS_MOS",");
+	if (vd->vdev_mg->mg_class[SPA_CLASS_DMU] != NULL)
+		(void) strcat(class, VDEV_CLASS_DMU",");
+	if (vd->vdev_mg->mg_class[SPA_CLASS_DDT] != NULL)
+		(void) strcat(class, VDEV_CLASS_DDT",");
+
+	if (strlen(class) == 1)
+		class[0] = '\0';
+	else
+		class[strlen(class) - 1] = ')';
+
+	(void) printf("\tvdev %10llu   %s\n\t%-10s%5llu   %-19s   %-15s   "
+	    "%-10s\n", (u_longlong_t)vd->vdev_id, class,
 	    "metaslabs", (u_longlong_t)vd->vdev_ms_count,
 	    "offset", "spacemap", "free");
 	(void) printf("\t%15s   %19s   %15s   %10s\n",
 	    "---------------", "-------------------",
 	    "---------------", "-------------");
 }
 
+/*
+ * dump dedicated classes
+ */
+static void
+dump_metaslab_class(metaslab_class_t *mc, const char *name)
+{
+	uint64_t fragmentation;
+	char vdevname[16];
+
+	if (metaslab_class_get_space(mc) == 0)
+		return;
+
+	(void) snprintf(vdevname, sizeof (vdevname), "%s vdev", name);
+
+	metaslab_class_histogram_verify(mc);
+
+	(void) printf("\t%s class\tfragmentation", name);
+	fragmentation = metaslab_class_fragmentation(mc);
+	if (fragmentation == ZFS_FRAG_INVALID)
+		(void) printf("\t%3s\n", "-");
+	else
+		(void) printf("\t%3llu%%\n", (u_longlong_t)fragmentation);
+
+	dump_histogram(mc->mc_histogram, RANGE_TREE_HISTOGRAM_SIZE, 0);
+}
+
+/*
+ * build class string in alphabetical order: "ddt,dmu,log,mos"
+ */
+static int
+vdev_class_string(vdev_t *vd, boolean_t secondary, char *bufptr, size_t buflen)
+{
+	if (buflen < 16)
+		return (EINVAL);
+
+	if (vd->vdev_mg == NULL)
+		return (ENODATA);
+
+	if (vd->vdev_mg->mg_class[SPA_CLASS_GENERAL] != NULL)
+		return (ENODATA);
+
+	bufptr[0] = '\0';
+
+	if (vd->vdev_mg->mg_class[SPA_CLASS_DDT] != NULL) {
+		(void) strncat(bufptr, VDEV_CLASS_DDT",", buflen);
+		buflen -= sizeof (VDEV_CLASS_DDT);
```

Comment:
This is decrementing correctly because sizeof(VDEV_CLASS_DDT) includes the trailing NUL, which makes up for the comma that you are not counting, right?